timemachine/cpp/src/streamed_potential_runner.cu

```diff
@@ -0,0 +1,76 @@
+#include "gpu_utils.cuh"
+#include "kernel_utils.cuh"
+#include "streamed_potential_runner.hpp"
+
+namespace timemachine {
+
+StreamedPotentialRunner::StreamedPotentialRunner() : streams_(0), events_(0) {
+    // Setup the event that is used to sync with the incoming stream
+    gpuErrchk(cudaEventCreateWithFlags(&sync_event_, cudaEventDisableTiming));
+};
+
+StreamedPotentialRunner::~StreamedPotentialRunner() {
+    for (int i = 0; i < streams_.size(); i++) {
+        gpuErrchk(cudaStreamDestroy(streams_[i]))
+    }
+    for (int i = 0; i < events_.size(); i++) {
+        gpuErrchk(cudaEventDestroy(events_[i]))
+    }
+    gpuErrchk(cudaEventDestroy(sync_event_));
+}
+
+cudaStream_t StreamedPotentialRunner::_get_potential_stream(int index) {
+    auto num_streams = streams_.size();
+    if (index < num_streams) {
+        return streams_[index];
+    }
+    // Expect stream to be the next increment
+    if (num_streams != index) {
+        throw std::runtime_error("Asked for new index out of order");
+    }
+    cudaStream_t new_stream;
+    // Create stream that doesn't block with the null stream
+    gpuErrchk(cudaStreamCreateWithFlags(&new_stream, cudaStreamNonBlocking));
+
+    cudaEvent_t new_event;
+    // Create stream with timings disabled
+    gpuErrchk(cudaEventCreateWithFlags(&new_event, cudaEventDisableTiming));
+    streams_.push_back(new_stream);
+    events_.push_back(new_event);
+    return new_stream;
+}
+
+cudaEvent_t StreamedPotentialRunner::_get_potential_event(int index) {
+    auto num_events = events_.size();
+    if (index >= num_events) {
+        throw std::runtime_error("No event with index " + std::to_string(index));
+    }
+    return events_[index];
+}
+
+// wrap execute_device
+void StreamedPotentialRunner::execute_potentials(
+    std::vector<std::shared_ptr<BoundPotential>> &bps,
+    const int N,
+    const double *d_x,
+    const double *d_box,
+    unsigned long long *d_du_dx,
+    unsigned long long *d_du_dp,
+    unsigned long long *d_u,
+    cudaStream_t stream) {
+    // Always sync the new streams with the incoming stream to ensure that the state
+    // of the incoming buffers is valid
+    gpuErrchk(cudaEventRecord(sync_event_, stream));
+    for (int i = 0; i < bps.size(); i++) {
+        cudaStream_t pot_stream = this->_get_potential_stream(i);
+
+        gpuErrchk(cudaStreamWaitEvent(pot_stream, sync_event_));
+        bps[i]->execute_device(N, d_x, d_box, d_du_dx, d_du_dp, d_u, pot_stream);
+        cudaEvent_t event = this->_get_potential_event(i);
+        // Tell the main stream to synchronize on all of the potential streams
+        gpuErrchk(cudaEventRecord(event, pot_stream));
+        gpuErrchk(cudaStreamWaitEvent(stream, event));
+    }
+};
```

Comment:
This is not a potential, this is just a way to run a series of bound potentials. 

Maybe StreamedBoundPotentialRunner?

Reply:
Ah, I guess one difference is that this only accepts bound potentials. Otherwise, though, isn't it basically a summed potential (i.e. takes a list of potentials and computes their sum given (coords, box) arguments)?

Maybe this suggestion would make more sense if the interface allowed passing unbound potentials and parameters, similar to the existing bound potential. As it is, is it  basically a `ParallelizedBoundSummedPotential`? (that's a mouthful)

Reply:
It is similar to a summed potential in the outcome, though this doesn't maintain any parameters. This API acts on any arbitrary set of BoundPotentials.

Having to construct this class around a set of bound potentials would result in creating and destroying many of these objects as the number of potentials passed to the integrator/barostat changes (local vs global md). Or some clever way of tracking which would seem like unnecessary complexity.

The solution that comes to mind is to split this runner up into two APIs:
* Syncing class that provides streams, events with APIs for syncing less verbosely
* ParallelizedBoundPotentialRunner for running a series of bound potentials in separate streams that relies on the syncing class

Then use the syncing class to duplicate the parallelism in the Summed/Fanout potential. How does that sound?

Reply:
> Having to construct this class around a set of bound potentials would result in creating and destroying many of these objects as the number of potentials passed to the integrator/barostat changes

Makes sense; I hadn't considered that the set of potentials passed to the integrator changes dynamically in local MD.

> The solution that comes to mind is to split this runner up into two APIs

This sounds reasonable to me.

> Then use the syncing class to duplicate the parallelism in the Summed/Fanout potential

My only concern is if we want to maintain sequential versions of these (in case there are potentials we want to explicitly avoid running in parallel), this would require added classes and complexity. Maybe better to split this up into a separate PR?

Reply:
> My only concern is if we want to maintain sequential versions of these (in case there are potentials we want to explicitly avoid running in parallel), this would require added classes and complexity. Maybe better to split this up into a separate PR?

I was going to add a `serial` flag to the syncing class, so parallelism could be disabled. Though I think all potentials should be capable of being run in parallel, but might be missing something.

Reply:
Modified Summed potentials to be parallelized in https://github.com/proteneer/timemachine/pull/1024/commits/1ee3b0f64d0f5801675a078ae4852e88476edd07